Various improvements

marmarek opened the pull request:
- qvm-shutdown fixes
- qvm-firewall API update (depends on https://github.com/QubesOS/qubes-core-admin/pull/119)
- two-stage volume clone (depends on https://github.com/QubesOS/qubes-core-admin/pull/120)
- do not try to clone `uuid` property
- qvm-template-postprocess fixes

marmarek committed:
events: fix calling mgmt.Events from VM

asyncio.create_subprocess_exec expects program and arguments directly,
not as a list.

marmarek committed:
tools/qvm-template-postprocess: improve error handling

Don't fail the whole process when "just" appmenus import fails.
But if data import fails, remove the VM

Also update for vm.run_service_for_stdio raising CalledProcessError.

marmarek committed:
tools: move event loop creation/closing to main function

Do not close event loop in utility function - handle it only in main().
For this reason, change appropriate functions to coroutines.

Fixes QubesOS/qubes-issues#2865

marmarek committed:
tests: handle returning different response for the same request

Allow programming different responses for the same request when called
multiple times. This is useful for example for shutdown tests - first
domain is running, but after issuing shutdown request is is not.

marmarek committed:
storage: rework clone as two-stage operation

Split clone to two Admin API calls - one to the source volume, then
other to destination.

marmarek committed:
app: do not clone 'uuid' property

Cloned VM have new UUID

marmarek committed:
firewall: drop GetPolicy/SetPolicy calls

Firewall policy is now hardcoded to 'drop'. Keep the property, so anyone
trying to assign it will get an exception

QubesOS/qubes-issues#2869

marmarek committed:
tools: remove policy handling from qvm-firewall tool

Follow the API removal

QubesOS/qubes-issues#2869

marmarek committed:
tools: split calling qubes.PostInstall service to a separate function

The code is long enough to warrant separate function (suggested by
pylint).

marmarek committed:
tools: clarify help in qvm-template-postprocess

woju commented on qubesadmin/tools/qvm_shutdown.py:
```diff
@@ -48,37 +48,66 @@
-
+    # care about order only when --wait is used:
+    if args.wait:
+        domains_and_distance = [(vm, distance_from_netvm(vm))
```
I don't think this is right, since extensions can introduce other dependencies (like device attached for example) and disallow vm shutdown for that (or other) reason. This code assumes that the only dependency between vms is expressed as netvms.

I'd write it as (this is only a rough idea):
```python
@asyncio.coroutine
def shutdown_vm(vm):
  yield from vm.shutdown()
  return vm

for i in range(max_rounds): # maybe 7
  if not vms_to_shutdown:
    break
  done, pending = yield from asyncio.wait((shutdown_vm(vm) for vm in vms_to_shutdown), timeout=20)
  # (or whatever timeout is deemed OK)
  # TODO handle pending and don't call vm.shutdown() more than once
  for fut in done:
    try:
      vms_to_shutdown.remove(fut.result())
    except QubesException:  # if vm.shutdown() raises exception, it gets re-raised by fut.result()
      pass

# after max_rounds, they are fair game
if vms_to_shutdown:
  yield from asyncio.wait(vm.kill() for vm in vms_to_shutdown)
```

That way we allow for more flexible ordering.

marmarek replied:
It is more complex than this - if `vm.shutdown()` raise an exception, VM should go to the next round. But if timeout, it should be killed, at the end of the same round. Otherwise for example hanging `untrusted` VM would block attempt to clean shutdown `sys-net` and `sys-firewall`.

marmarek replied:
For now I've dropped qvm-shutdown rework from this PR, will work on it later. @woju please review other changes.

marmarek committed:
app: fix policy deny reporting when running in VM

qrexec-client-vm non-zero exit code means policy have denied the call.
Treat this exactly the same as empty response (in dom0 case).
